soc/xtensa/intel_adsp/ace_v1x/power.c

```diff
@@ -26,29 +26,194 @@
  * (each bit corresponds to one ebb)
  * @param response_to_ipc       flag if ipc response should be send during power down
  */
-extern void ace_power_down(bool disable_lpsram, uint32_t *hpsram_pg_mask,
+extern void power_down(bool disable_lpsram, uint32_t *hpsram_pg_mask,
 			   bool response_to_ipc);
 
+static ALWAYS_INLINE uint32_t read_intenable(void)
+{
+	uint32_t intenable;
+
+	__asm__ volatile("rsr.intenable %0" : "=r"(intenable));
+	return intenable;
+}
+
+static ALWAYS_INLINE uint32_t read_threadptr(void)
+{
+	uint32_t threadptr;
+
+	__asm__ volatile(
+		"rur.threadptr %0" : "=a"(threadptr));
+	return threadptr;
+}
+
+static ALWAYS_INLINE void write_threadptr(uint32_t threadptr)
+{
+	__asm__ volatile(
+		"wur.threadptr %0" : : "a" (threadptr) : "memory");
+}
+
+static ALWAYS_INLINE uint32_t read_vecbase(void)
+{
+	uint32_t vecbase;
+
+	__asm__ volatile("rsr.vecbase %0"
+		: "=a"(vecbase));
+	return vecbase;
+}
+
+static ALWAYS_INLINE void write_vecbase(uint32_t vecbase)
+{
+	__asm__ volatile("wsr.vecbase %0"
+		: : "a"(vecbase) : "memory");
+}
+
+static ALWAYS_INLINE uint32_t read_excsave2(void)
+{
+	uint32_t excsave2;
+
+	__asm__ __volatile__("rsr.excsave2 %0"
+		: "=a"(excsave2));
+	return excsave2;
+}
+
+static ALWAYS_INLINE void write_excsave2(uint32_t excsave2)
+{
+	__asm__ __volatile__("wsr.excsave2 %0"
+		: : "a"(excsave2) : "memory");
+}
+
+static ALWAYS_INLINE uint32_t read_excsave3(void)
+{
+	uint32_t excsave3;
+
+	__asm__ __volatile__("rsr.excsave3 %0"
+		: "=a"(excsave3));
+	return excsave3;
+}
+
+static ALWAYS_INLINE void write_excsave3(uint32_t excsave3)
+{
+	__asm__ __volatile__("wsr.excsave3 %0"
+		: : "a"(excsave3) : "memory");
+}
+
+#define BITMASK_FROM_RANGE_UINT32(from, to) ((0xFFFFFFFF << (from)) & \
+					    (0xFFFFFFFF >> (sizeof(uint32_t) * 8 - 1 - (to))))
+#define LPSCTL_BATTR_MASK                   BITMASK_FROM_RANGE_UINT32(12, 16)
+#define LPSRAM_MAGIC_VALUE 0x13579BDF
+
+struct lpsram_header {
+	uint32_t alt_reset_vector;
+	uint32_t adsp_lpsram_magic;
+	void *lp_restore_vector;
+	uint32_t reserved;
+	uint32_t slave_core_vector;
+	uint8_t rom_bypass_vectors_reserved[0xC00 - 0x14];
+};
+
+enum PowerFlow {
+	D0_TO_D0i3,
+	D0i3_TO_D0
+};
```

Comment:
That looks really weird, can't you map the flow in relation to Zephyr pm states ? Or even better, just use `enum pm_state` ?